src/crimson/os/seastore/async_cleaner.h

```diff
@@ -1161,10 +1162,15 @@ class AsyncCleaner : public SegmentProvider, public Trimmer {
       return seastar::do_until(
 	[this] {
 	  cleaner.log_gc_state("GCProcess::run_until_halt");
-	  return !cleaner.gc_should_run();
+	  return !cleaner.gc_should_run() || is_running_until_halt;
 	},
 	[this] {
-	  return cleaner.do_gc_cycle();
+	  is_running_until_halt = true;
+	  return cleaner.do_gc_cycle(
+	  ).then([this]() {
+	    is_running_until_halt = false;
+	    return seastar::now();
+	  });
 	});
     }
```

Comment:
I think an important consideration here is to make sure that checking and setting `is_running_until_halt` are atomic inside the same continuation. And same applies to resetting `is_running_until_halt` and breaking the loop.

Does this look better?
```c++
gc_cycle_ret run_until_halt() {
  ceph_assert(is_stopping());
  if (is_running_until_halt) {
    return seastar::now();
  }
  is_running_until_halt = true;
  return seastar::do_until(
    [this] {
      cleaner.log_gc_state("GCProcess::run_until_halt");
      assert(is_running_until_halt);
      if (cleaner.gc_should_run()) {
        return false;
      } else {
        is_running_until_halt = false;
        return true;
      }
    },
    [this] {
      return cleaner.do_gc_cycle();
    }
  );
}
```